File: algorithms/java/src/test/java/leetcode/LC100SameTree.java
package leetcode.lc100_same_tree;

import java.util.ArrayList;

import org.junit.jupiter.api.Test;


import static org.junit.jupiter.api.Assertions.assertEquals;

/**
 * https://leetcode.com/problems/same-tree/
 * #easy
 */
public final class LC100SameTree {
    public static class TreeNode {
        public int val;
        public TreeNode left;
        public TreeNode right;

        public TreeNode(final int x) {
            val = x;
        }
    }

    public final class Solution {
        public boolean isSameTree(final TreeNode tree1, final TreeNode tree2) {
            if (tree1 == null && tree2 == null) return true;
            if (tree1 == null || tree2 == null) return false;
            ArrayList<TreeNode> queue1 = new ArrayList<>();
            ArrayList<TreeNode> queue2 = new ArrayList<>();
            queue1.add(tree1);
            queue2.add(tree2);
            while (queue1.size() > 0 && queue2.size() > 0) {
                TreeNode node1 = queue1.remove(0);
                TreeNode node2 = queue2.remove(0);
                if (node1.val != node2.val) return false;
                if (node1.left != null && node2.left != null) {
                    queue1.add(node1.left);
                    queue2.add(node2.left);
                } else if (node1.left != null || node2.left != null) return false;
                if (node1.right != null && node2.right != null) {
                    queue1.add(node1.right);
                    queue2.add(node2.right);
                } else if (node1.right != null || node2.right != null) return false;
            }
            return queue1.size() == queue2.size();
        }
    }

    @Test
    public void test_different() throws Exception {
        TreeNode an0 = new TreeNode(0);
        TreeNode bn0 = new TreeNode(0);
        TreeNode an1 = new TreeNode(1);
        TreeNode bn1 = new TreeNode(1);
        TreeNode an2 = new TreeNode(2);
        TreeNode bn2 = new TreeNode(3);
        an0.left = an1;
        an0.right = an2;
        bn0.left = bn1;
        bn0.right = bn2;
        an1.left = null;
        an2.right = null;
        bn1.left = null;
        bn2.right = null;
        assertEquals(false, new Solution().isSameTree(an0, bn0));
    }

    @Test
    public void test_same() throws Exception {
        TreeNode an0 = new TreeNode(0);
        TreeNode bn0 = new TreeNode(0);
        TreeNode an1 = new TreeNode(1);
        TreeNode bn1 = new TreeNode(1);
        TreeNode an2 = new TreeNode(2);
        TreeNode bn2 = new TreeNode(2);
        an0.left = an1;
        an0.right = an2;
        bn0.left = bn1;
        bn0.right = bn2;
        an1.left = null;
        an2.right = null;
        bn1.left = null;
        bn2.right = null;
        assertEquals(true, new Solution().isSameTree(an0, bn0));
    }
}

File: algorithms/code/leetcode/lc100_same_tree/lc100_same_tree.py
# https://leetcode.com/problems/same-tree/

from __future__ import annotations

import unittest


class TreeNode:
    def __init__(
        self,
        val: int = 0,
        left: 'TreeNode | None' = None,
        right: 'TreeNode | None' = None,
    ) -> None:
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def isSameTree(
        self, tree1: TreeNode | None, tree2: TreeNode | None
    ) -> bool:
        if tree1 is None or tree2 is None:
            return tree1 is None and tree2 is None
        queue: list[tuple[TreeNode, TreeNode]] = []
        queue.append((tree1, tree2))
        while queue:
            node1, node2 = queue.pop(0)
            if node1.val != node2.val:
                return False
            node1_left, node2_left = node1.left, node2.left
            if node1_left is not None and node2_left is not None:
                queue.append((node1_left, node2_left))
            elif not (node1_left is None and node2_left is None):
                return False
            node1_right, node2_right = node1.right, node2.right
            if node1_right is not None and node2_right is not None:
                queue.append((node1_right, node2_right))
            elif not (node1_right is None and node2_right is None):
                return False
        return True


class TestCode(unittest.TestCase):
    def test_different(self) -> None:
        an0 = TreeNode(0)
        bn0 = TreeNode(0)
        an1 = TreeNode(1)
        bn1 = TreeNode(1)
        an2 = TreeNode(2)
        bn2 = TreeNode(3)
        an0.left = an1
        an0.right = an2
        bn0.left = bn1
        bn0.right = bn2
        an1.left = None
        an2.right = None
        bn1.left = None
        bn2.right = None
        assert not Solution().isSameTree(an0, bn0)

    def test_same(self) -> None:
        an0 = TreeNode(0)
        bn0 = TreeNode(0)
        an1 = TreeNode(1)
        bn1 = TreeNode(1)
        an2 = TreeNode(2)
        bn2 = TreeNode(2)
        an0.left = an1
        an0.right = an2
        bn0.left = bn1
        bn0.right = bn2
        an1.left = None
        an2.right = None
        bn1.left = None
        bn2.right = None
        assert Solution().isSameTree(an0, bn0)

    def test_one_empty(self) -> None:
        tree = TreeNode(0)
        assert not Solution().isSameTree(tree, None)
        assert not Solution().isSameTree(None, tree)

    def test_both_empty(self) -> None:
        assert Solution().isSameTree(None, None)

    def test_left(self) -> None:
        an0 = TreeNode(0)
        bn0 = TreeNode(0)
        an1 = TreeNode(1)
        an0.left = an1
        assert not Solution().isSameTree(an0, bn0)

    def test_right(self) -> None:
        an0 = TreeNode(0)
        bn0 = TreeNode(0)
        an1 = TreeNode(1)
        an0.right = an1
        assert not Solution().isSameTree(an0, bn0)
